dttools/src/jx_match.c

```diff
@@ -0,0 +1,198 @@
+/*
+Copyright (C) 2016- The University of Notre Dame
+This software is distributed under the GNU General Public License.
+See the file COPYING for details.
+*/
+
+#include <stdarg.h>
+#include <string.h>
+
+#include "jx.h"
+#include "jx_match.h"
+
+void *jx_match_null(struct jx *j) {
+	return jx_istype(j, JX_NULL) ? (void *) j : NULL;
+}
+
+int *jx_match_boolean(struct jx *j, int *v) {
+	if (jx_istype(j, JX_BOOLEAN)) {
+		if (v) {
+			*v = !!j->u.boolean_value;
+		}
+		return &j->u.boolean_value;
+	} else {
+		return NULL;
+	}
+}
+
+jx_int_t *jx_match_integer(struct jx *j, jx_int_t *v) {
+	if (jx_istype(j, JX_INTEGER)) {
+		if (v) {
+			*v = j->u.integer_value;
+		}
+		return &j->u.integer_value;
+	} else {
+		return NULL;
+	}
+}
+
+double *jx_match_double(struct jx *j, double *v) {
+	if (jx_istype(j, JX_DOUBLE)) {
+		if (v) {
+			*v = j->u.double_value;
+		}
+		return &j->u.double_value;
+	} else {
+		return NULL;
+	}
+}
+
+char **jx_match_string(struct jx *j, char **v) {
+	if (jx_istype(j, JX_STRING)) {
+		if (v) {
+			*v = strdup(j->u.string_value);
```

Comment:
Here, as per our discussion, I think if the strdup fails we should return NULL. That way if the match is successful we know that v is also usable if requested. This can be bypassed by setting v to NULL as a parameter in the case were you just want a point.